### PR TITLE
Optimize HashPMap.get and other operations: do not create unnecessary iterator instances

### PR DESCRIPTION
**Pull request by udalov**
We use PCollections as an internal cache for classes in Kotlin reflection library ([proof](https://github.com/JetBrains/kotlin/blob/master/core/reflection.jvm/src/kotlin/reflect/jvm/internal/kClassCache.kt)) and have recently found a problem: `HashPMap.get` seems to create lots of unnecessary `ConsPStack$1` objects, which are `ConsPStack` iterators, even on simple `get` access! Here's the original report (never mind my misleading answer there):
http://stackoverflow.com/questions/34214290/kotlin-massive-amounts-of-conspstack-how-can-i-avoid/34214718#34214718

This pull request optimizes iteration over `ConsPStack` inside `HashPMap`. First, I've made sure only `ConsPStack` is used as the `PSequence` implementation by making the types of parameters and fields more specific. In the second commit, I've replaced the for loop with the manual iteration over the `ConsPStack` instance. In the last commit, I've made anonymous classes inner with meaningful names, so that one will observe a more insightful `ConsPStack$Itr` in the stack trace instead of `ConsPStack$1`.

### CODE REVIEW

File: src/main/java/org/pcollections/HashPMap.java [hunk 1]
```diff
@@ -134,28 +139,29 @@ public V get(final Object key) {
...
 	}
 
 	
 //// PRIVATE STATIC UTILITIES ////
-	private static <K,V> int keyIndexIn(final PSequence<Entry<K,V>> entries, final Object key) {
+	private static <K,V> int keyIndexIn(final ConsPStack<Entry<K,V>> entries, final Object key) {
 		int i=0;
-		for(Entry<K,V> entry : entries) {
+		while(entries != null && !entries.isEmpty()) {
+			Entry<K,V> entry = entries.first;
```
Comment (ilya-g) [inline src/main/java/org/pcollections/HashPMap.java]: This seems to be an infinite loop. `entries` is immutable and final, so the condition of loop never changes.

Comment (udalov) [inline src/main/java/org/pcollections/HashPMap.java]: Thanks, fixed

**Comment by bowbahdoe** [comment]
Any update on this?

File: src/main/java/org/pcollections/ConsPStack.java [hunk 2]
```diff
@@ -84,38 +84,46 @@ public int size() {
...
-				E e = next.first;
-				next = next.rest;
-				return e;
-			}
-			public E previous() {
-				System.err.println("ConsPStack.listIterator().previous() is inefficient, don't use it!");
-				next = subList(index-1); // go from beginning...
-				return next.first;
-			}
+	private class Itr implements ListIterator<E> {
```
Comment (hrldcpr) [inline src/main/java/org/pcollections/ConsPStack.java:90]: should this be static?

File: src/main/java/org/pcollections/HashPMap.java [hunk 3]
```diff
@@ -35,15 +35,16 @@
 	 * @return a map backed by an empty version of intMap,
 	 * 	i.e. backed by intMap.minusAll(intMap.keySet())
 	 */
+	@SuppressWarnings("unchecked")
 	public static <K,V> HashPMap<K,V> empty(final PMap<Integer,PSequence<Entry<K,V>>> intMap) {
-		return new HashPMap<K,V>(intMap.minusAll(intMap.keySet()), 0); }
+		return new HashPMap<K,V>((PMap) intMap.minusAll(intMap.keySet()), 0); }
```
Comment (hrldcpr) [inline src/main/java/org/pcollections/HashPMap.java:40]: why is this case necessary?

File: src/main/java/org/pcollections/HashPMap.java [hunk 4]
```diff
@@ -53,27 +54,28 @@ private HashPMap(final PMap<Integer,PSequence<Entry<K,V>>> intMap, final int siz
...
-					V value = get(((Entry<?,?>)e).getKey());
-					return value!=null && value.equals(((Entry<?,?>)e).getValue());
-				}
-			};
+			entrySet = new EntrySet();
 		return entrySet;
 	}
 
-	
+	private class EntrySet extends AbstractSet<Entry<K,V>> {
```
Comment (hrldcpr) [inline src/main/java/org/pcollections/HashPMap.java:61]: static?

**Pull request closed without merging**